clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java

```diff
@@ -3373,19 +3376,45 @@ protected void processMessageAndMaybeProduceToKafka(
     // Write to views
     if (hasViewWriters()) {
       Put newPut = writeComputeResultWrapper.getNewPut();
-      // keys will be serialized with chunk suffix during pass-through mode in L/F NR if chunking is enabled
-      boolean isChunkedKey = isChunked() && !partitionConsumptionState.isEndOfPushReceived();
+      Map<String, Set<Integer>> viewPartitionMap = null;
+      if (!partitionConsumptionState.isEndOfPushReceived()) {
+        // NR pass-through records are expected to carry view partition map in the message header
+        viewPartitionMap = extractViewPartitionMap(consumerRecord.getPubSubMessageHeaders());
+      }
       Lazy<GenericRecord> newValueProvider = writeComputeResultWrapper.getValueProvider();
       queueUpVersionTopicWritesWithViewWriters(
           partitionConsumptionState,
-          (viewWriter) -> viewWriter
-              .processRecord(newPut.putValue, keyBytes, newPut.schemaId, isChunkedKey, newValueProvider),
+          (viewWriter, viewPartitionSet) -> viewWriter
+              .processRecord(newPut.putValue, keyBytes, newPut.schemaId, viewPartitionSet, newValueProvider),
+          viewPartitionMap,
           produceToVersionTopic);
     } else {
       produceToVersionTopic.run();
     }
   }
 
+  protected Map<String, Set<Integer>> extractViewPartitionMap(PubSubMessageHeaders pubSubMessageHeaders) {
+    Map<String, Set<Integer>> viewPartitionMap = null;
+    for (PubSubMessageHeader header: pubSubMessageHeaders) {
```

Comment:
Do we really have to iterate through these? PubsubMessageHeaders already has a map in it.  Couldn't we just ask for that header specifically?